### PR TITLE
Long Animation Frame (LoAF) explainer

### PR DESCRIPTION
**Pull request by noamr**
Add explainer for long-task revamp

See https://github.com/w3c/event-timing/issues/122 and https://github.com/w3c/longtasks/issues/89

### CODE REVIEW

**Review by mmocny — COMMENTED** [review]
Exciting!

File: loaf-explainer.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+All of the above are part of the same issue - a task is an incomplete and inaccurate cadence to measure main-thread blocking. It's either too granular (as several tasks together may be the cause of blocking) or too coarse (as it may batch together several event handlers and so on).
+
+## Introducing LoAF
+
+LoAF (long animation frame) is a new proposed cadence (period of time) and performance entry type, meant to be a progression of the long task concept.
+
+It's the time measured between when the main thread started doing any work (see `startTime` [here](https://html.spec.whatwg.org/#event-loop-processing-model)), until it is [ready to paint](https://html.spec.whatwg.org/#event-loop-processing-model:mark-paint-timing). It may include several tasks (e.g. a few short ones and then a long one). Because it ends at the paint-mark time, it
+includes all the rendering obsevers (requestAnimationFrame, ResizeObserver etc.) and
+may or may not include presentation time (as that is somewhat implementation specific).
```
[the remark this answers is not in the document]
Comment (noamr) [reply loaf-explainer.md:39]: Yes, I'm thinking that the `presentationTime` here would not affect `duration`, but we can expose that information nontheless.

File: loaf-explainer.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    // getComputedStyle(). This is only taken into account if there is also a layout/style update
+    // in the final rendering phase.
+    totalForcedStyleAndLayoutDuration,
+
+    // The number of times there was a render opportunity but the main thread was busy
+    // processing this frame
+    discardedRenderOpportunities,
+
+    // Whether this long frame was blocking input/animation in practice
+    blocking: 'ui-event' | 'animation' | 'none',
```
[the remark this answers is not in the document]
Comment (noamr) [reply loaf-explainer.md:70]: Yes, I think in that case we'd say we're blocking a UI event.

File: loaf-explainer.md [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+- Is the frame blocking animation/input-feedback *in practice*. Note that a frame that blocks actual UI events would also be accessible via [event timing](https://w3c.github.io/event-timing/).
+- User scripts processed during the time of the frame, be it callbacks, event handlers, promise resolvers, or script block parsing and evaluation, and information about those scripts (source, how long they've been delayed for).
+
+## How a LoAF entry might look like
+```js
+const someLongAnimationFrameEntry = {
+    entryType: "long-animation-frame",
+
+    // https://html.spec.whatwg.org/#event-loop-processing-model, see |taskStartTime|
+    startTime: taskStartTime,
```
[the remark this answers is not in the document]
Comment (noamr) [reply loaf-explainer.md:54]: Actually I changed this - it's wrong. It should be the first task after the last "update the rendering" that had a rendering opportunity.

File: loaf-explainer.md [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+It's the time measured between when the main thread started doing any work (see `startTime` [here](https://html.spec.whatwg.org/#event-loop-processing-model)), until it is [ready to paint](https://html.spec.whatwg.org/#event-loop-processing-model:mark-paint-timing). It may include several tasks (e.g. a few short ones and then a long one). Because it ends at the paint-mark time, it
+includes all the rendering obsevers (requestAnimationFrame, ResizeObserver etc.) and
+may or may not include presentation time (as that is somewhat implementation specific).
+
+In addition to making the cadence fit better with what it measures, the entry would include extra information to help understand what made it long, and what kind of consequences it had:
+
+- Time spent in forced layout/style calculations - e.g. calling `getBoundingClientRect`, doing more processing, and then rendering (also known as "layout thrashing" or "forced reflow").
+- How many rendering opportuinities were missed - counting points in time where the browser was ready to receive a new paint but the main thread was busy with this LoAF.
```
[the remark this answers is not in the document]
Comment (noamr) [reply loaf-explainer.md:44]: Removed for now, would rather concentrate on concensus.

File: loaf-explainer.md [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        if (totalRenderingOpportunities > 3 || performance.now() - startTime > 50)
+            reportLongAnimationFrame();
+        render();
+    }
+}
+```
+
+### Some notes about processing
+
+1. relying on "discarded rendering opportunities" as the qualifier for sluggishness alongside (or
```
[the remark this answers is not in the document]
Comment (noamr) [reply loaf-explainer.md]: Rendering priority is an unspecified implementation-specific thing.
I think at least to begin with I'd keep it well lit and only measure active visible documents that remain that way throughout the frame.

File: loaf-explainer.md [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+}
+```
+
+### Security & Privacy Considerations
+
+At the most part, LoAF only exposes information across same-origin windows. Information about
+scripts within a window is already observable, e.g. using resource timing or a service worker.
+
+However, LoAF might expose rendering information for a particular document tree that may be
+cross-origin (same-agent/site). The details about rendering the frame, such as
```
Comment (yoavweiss) [inline loaf-explainer.md:293]: Is this statement relying on cross-site isolation?

File: loaf-explainer.md [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+`styleAndLayoutStartTime`, are proposed to be visible to all the same-agent windows that are
+rendered serially. That's because this information is already observable, by using
+`requestAnimationFrame` and `ResizeObserver` and measuring the delay between them. The premise is
+that global "update the rendering" timing information is already observable across same-agent
+windows, so exposing it directly does not leak new cross-origin information. However, the idea
+exposing less information to cross-origin same-agent subframes (as in, expose ) is open for
```
Comment (yoavweiss) [inline loaf-explainer.md:299]: missing word after "expose "?

File: loaf-explainer.md (hunk 6 above)
Comment (yoavweiss) [reply loaf-explainer.md:293]: Given that the rest of the paragraph talks about agents, maybe drop the "site" bit?

File: loaf-explainer.md [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+
+    1. An observer fires only if its rendering was blocked by the long frame in practice, or if
+    the long task (that didn't cause a render) belonged to that page.
+
+    1. Breakdown to scripts is only available to the frame where they were invoked. Other frames
+    receive an "opaque" breakdown: attribution of a blocking task to a different window - similar to
+    the existing [attribution](https://w3c.github.io/longtasks/#sec-TaskAttributionTiming).
+
+
+1. To avoid the magic 50hz number, consider making the threshold configurable,
```
Comment (yoavweiss) [inline loaf-explainer.md:318]: 50ms?

**Review by yoavweiss — APPROVED** [review]
A few nits but otherwise LGTM

**Pull request merged**